File: src/snapshot.js
const os = require('node:os');
const path = require('node:path');
const { execFile } = require('node:child_process');

const SNAPSHOT_MAX_BUFFER = 8 * 1024 * 1024;

function readSnapshot(binding, { registry = path.join(os.homedir(), '.agents/work-control/pr-lanes.json'),
  ladderDir = path.join(os.homedir(), '.claude/skills/conduct-status/scripts'), now, signal } = {}) {
  const args = [path.join(__dirname, 'snapshot.py'), '--registry', registry, '--ladder-dir', ladderDir];
  if (now !== undefined) args.push('--now', String(now));
  return new Promise(resolve => {
    const child = execFile('/usr/bin/python3', args, { timeout: 3000, maxBuffer: SNAPSHOT_MAX_BUFFER, signal }, (error, stdout) => {
      try {
        const result = JSON.parse(stdout);
        resolve(error ? { unavailable: result.unavailable || 'snapshot reader failed' } : result);
      } catch { resolve({ unavailable: error?.message || 'invalid snapshot output' }); }
    });
    child.stdin.on('error', () => {});
    child.stdin.end(JSON.stringify(binding));
  });
}

function discordText(value) {
  return String(value).replace(/([\\`*_~>|])/g, '\\$1').replace(/@/g, '@\u200b');
}

function truncate(text, limit) {
  if (text.length <= limit) return text;
  if (limit <= 1) return text.slice(0, limit);
  return `${text.slice(0, limit - 1)}…`;
}

function renderSnapshot(snapshot) {
  if (snapshot.unavailable) return 'Automatic artifact snapshot: source unavailable.';
  const { context, lanes, source } = snapshot;
  const updated = Date.parse(context.updated || '');
  const stamp = Number.isFinite(updated) ? `<t:${Math.floor(updated / 1000)}:R>` : 'time not recorded';
  const lines = [`**Automatic artifact snapshot · ${discordText(snapshot.identity.provider)}**`,
    `Recorded context: ${context.freshness === 'current' ? 'as of source' : context.freshness} · ${stamp}`];
  const renderField = (label, field, list = false, limit = 400) => {
    if (field.state === 'invalid') return `${label}: invalid.`;
    if (field.state !== 'recorded') return `${label}: not recorded.`;
    if (list && !field.value.length) return `${label}: none recorded${context.freshness === 'stale' ? ' (stale source)' : ''}.`;
    const prefix = `${label}: `;
    if (!list) return `${prefix}${truncate(discordText(field.value), Math.max(1, limit - prefix.length))}`;
    let output = prefix;
    let included = 0;
    for (const value of field.value) {
      const item = discordText(typeof value === 'string' ? value : value.text);
      const separator = included ? ' / ' : '';
      if (output.length + separator.length + item.length > limit) break;
      output += separator + item;
      included += 1;
    }
    let omitted = field.value.length - included;
    if (!omitted) return output;
    let suffix = `${included ? ' / ' : ''}${omitted} item(s) omitted.`;
    if (!included) {
      const item = discordText(typeof field.value[0] === 'string' ? field.value[0] : field.value[0].text);
      const itemRoom = limit - output.length - suffix.length;
      if (itemRoom > 1) {
        output += truncate(item, itemRoom);
        included = 1;
        omitted -= 1;
        suffix = `${omitted ? ' / ' : ''}${omitted} item(s) omitted.`;
      }
    }
    if (!omitted) return output;
    if (output.length + suffix.length <= limit) return output + suffix;
    const room = Math.max(prefix.length, limit - suffix.length);
    output = output.slice(0, Math.max(prefix.length, room - 1)) + '…';
    return output + suffix;
  };
  lines.push(renderField('Intent', context.intent, false, 300), renderField('Owed by you', context.owedByOperator, true, 420),
    renderField('Owed to you', context.owedToOperator, true, 420), renderField('Next', context.next, true, 300));
  let omitted = Number.isInteger(snapshot.omittedLanes) ? Math.max(0, snapshot.omittedLanes) : 0;
  const footer = `Source revision ${String(source?.revision || 'unknown').slice(0, 10)}. Recorded claims, not a live process census.`;
  const omissionReserve = `${Number.MAX_SAFE_INTEGER} lane record(s) omitted from this preview.`.length;
  const canAdd = line => lines.join('\n').length + 1 + line.length + 1 + footer.length + 1 + omissionReserve <= 2000;
  for (const lane of lanes) {
    const line = `${discordText(lane.ticket || lane.id)}${lane.pr ? ` #${discordText(lane.pr)}` : ''}: ${lane.percent ?? '??'}%${lane.held ? ' held' : ''} · ${discordText(lane.stateNote || lane.phase || 'state not recorded')}\nNext: ${discordText(lane.next || 'not recorded')}`;
    if (canAdd(line)) lines.push(line);
    else omitted += 1;
  }
  if (!lanes.length) {
    const empty = 'No matching lane records. This does not establish completion.';
    if (canAdd(empty)) lines.push(empty);
  }
  if (omitted) lines.push(`${omitted} lane record(s) omitted from this preview.`);
  lines.push(footer);
  return lines.join('\n');
}

module.exports = { readSnapshot, renderSnapshot };

File: src/snapshot.py
#!/usr/bin/env python3
"""Read existing conductor records without entering a native session."""
import argparse
import datetime
import hashlib
import json
import pathlib
import sys

MAX_BYTES = 2 * 1024 * 1024
STALE_SECONDS = 30 * 60


def text(value, limit=600):
    return value if isinstance(value, str) and 0 < len(value) <= limit else None


def epoch(value):
    if not isinstance(value, str):
        return None
    try:
        stamp = datetime.datetime.fromisoformat(value.replace("Z", "+00:00"))
        return stamp.timestamp() if stamp.tzinfo else None
    except ValueError:
        return None


def field(record, key, kind):
    if key not in record:
        return {"state": "missing", "value": None}
    value = record[key]
    if kind == "text":
        valid = text(value) is not None
    elif kind == "steps":
        valid = isinstance(value, list) and len(value) <= 12 and all(text(v) for v in value)
    else:
        valid = isinstance(value, list) and len(value) <= 30
        if valid:
            valid = all(isinstance(v, dict) and text(v.get("id"), 128) and text(v.get("text"))
                        and epoch(v.get("since")) is not None for v in value)
        if valid:
            valid = len({v["id"] for v in value}) == len(value)
            value = [{k: v[k] for k in ("id", "text", "since")} for v in value]
    return {"state": "recorded" if valid else "invalid", "value": value if valid else None}


def snapshot(binding, registry, ladder_dir, now):
    sys.path.insert(0, str(ladder_dir))
    ladder_path = pathlib.Path(ladder_dir) / "lane_progress_ladder.py"
    ladder = {"__file__": str(ladder_path), "__name__": "lane_progress_ladder"}
    # Timestamp-based bytecode can miss equal-size edits within one second.
    exec(compile(ladder_path.read_bytes(), str(ladder_path), "exec"), ladder)
    PCT, canonical = ladder["PCT"], ladder["canonical"]

    source = pathlib.Path(registry)
    with source.open("rb") as stream:
        raw = stream.read(MAX_BYTES + 1)
    if len(raw) > MAX_BYTES:
        raise ValueError("registry exceeds snapshot read bound")
    data = json.loads(raw)
    if not isinstance(data, dict):
        raise ValueError("registry must be an object")
    owner = binding["conductorId"]
    records = data.get("_conductors", {})
    record = records.get(owner) if isinstance(records, dict) else None
    identity = {key: binding[key] for key in ("conductorId", "repoKey", "provider", "nativeId", "generation", "channelId")}
    status = "missing"
    if isinstance(record, dict):
        matches = (record.get("repository") == binding["repoKey"] and record.get("vendor") == binding["provider"]
                   and record.get("nativeId") == binding["nativeId"] and type(record.get("generation")) is int
                   and record["generation"] == binding["generation"])
        status = "recorded" if matches else "wrong-owner"
    if status != "recorded":
        record = {}
    updated = epoch(record.get("updated"))
    freshness = "unknown" if updated is None or updated > now else "stale" if now - updated >= STALE_SECONDS else "current"
    context = {
        "state": status, "updated": record.get("updated"), "freshness": freshness,
        "owedByOperator": field(record, "owed_by_operator", "owed"),
        "owedToOperator": field(record, "owed_to_operator", "owed"),
        "intent": field(record, "intent", "text"), "next": field(record, "next", "steps")
    }
    native_id = binding["nativeId"]
    owner_tokens = {owner, native_id, "session-" + binding["provider"] + "-" + native_id}
    lanes = []
    for key, lane in data.items():
        if key.startswith("_") or not isinstance(lane, dict) or lane.get("vendor") != binding["provider"]:
            continue
        repo = lane.get("repository", lane.get("repoKey"))
        lane_native_id = lane.get("nativeId", lane.get("native_id"))
        lane_generation = lane.get("generation")
        lane_identity_present = any(key in lane for key in ("nativeId", "native_id", "generation"))
        lane_proves_binding = (lane_native_id == native_id and type(lane_generation) is int
                               and lane_generation == binding["generation"])
        abbreviated_owner = (binding["provider"] == "claude" and
                             lane.get("conductor") == "session-claude-" + native_id[:8])
        if lane.get("conductor") not in owner_tokens and not (abbreviated_owner and lane_proves_binding):
            continue
        if lane_identity_present and not lane_proves_binding:
            continue
        if status != "recorded" and not lane_proves_binding:
            continue
        if repo is not None and repo != binding["repoKey"]:
            continue
        stage = canonical(lane.get("phase"))
        if stage == "done":
            continue
        held = stage in ("held", "parked", "frozen")
        percent = PCT.get(canonical(lane.get("phase_before_hold")) if held else stage)
        lanes.append({"id": key, "pr": lane.get("pr"), "ticket": text(lane.get("ticket"), 128),
                      "head": text(lane.get("head"), 128), "phase": text(lane.get("phase"), 200),
                      "percent": percent, "held": held, "next": text(lane.get("next")),
                      "stateNote": text(lane.get("state_note")), "updated": text(lane.get("updated"), 100)})
    lanes.sort(key=lambda row: row["id"])
    result = {"identity": identity, "context": context, "lanes": lanes[:30], "omittedLanes": max(0, len(lanes)-30)}
    result["id"] = hashlib.sha256(json.dumps(result, sort_keys=True).encode()).hexdigest()
    result["source"] = {"path": str(source.resolve()), "revision": hashlib.sha256(raw).hexdigest(),
                        "observedAt": datetime.datetime.fromtimestamp(now, datetime.timezone.utc).isoformat(),
                        "contextUpdated": record.get("updated")}
    if freshness == "current":
        result["expiresAt"] = updated + STALE_SECONDS
    elif freshness == "unknown" and updated is not None and updated > now:
        result["expiresAt"] = updated
    else:
        result["expiresAt"] = None
    return result


if __name__ == "__main__":
    parser = argparse.ArgumentParser()
    parser.add_argument("--registry", required=True)
    parser.add_argument("--ladder-dir", required=True)
    parser.add_argument("--now", type=float)
    args = parser.parse_args()
    try:
        binding = json.loads(sys.stdin.read(16000))
        now = args.now if args.now is not None else datetime.datetime.now(datetime.timezone.utc).timestamp()
        print(json.dumps(snapshot(binding, args.registry, args.ladder_dir, now), ensure_ascii=False, separators=(',', ':')))
    except (OSError, ValueError, KeyError, ImportError) as error:
        print(json.dumps({"unavailable": str(error)}))
        sys.exit(1)
